File: editor-server/src/routes/api/checkToken.ts
import type { Request, Response } from "express";

import redis from "../../redis";
import prisma from "../../prisma";

const checkToken = async (req: Request, res: Response) => {
  // check if cookie exists
  const { token } = req.cookies;
  console.log("checkToken", token);

  if (token) {
    // check if sessionid is valid
    const id = parseInt((await redis.get(token)) ?? "0");
    if (id) {
      // send back user data
      const user = await prisma.user.findUnique({
        where: {
          id,
        },
      });
      if (user) {
        res.send({ token });
      } else {
        res.status(404).send({ err: "User not found." });
      }
    }
  } else {
    res.status(401).send({ err: "Unauthorized." });
  }
};

export default checkToken;

File: editor-server/src/routes/api/check_token.rs
use crate::utils::authentication::verify_token;

use axum::{http::StatusCode, response::Json};
use axum_extra::extract::CookieJar;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize)]
pub struct CheckTokenResponse {
    token: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct CheckTokenFailedResponse {
    err: String,
}

/// check token endpoint
/// get token from cookie and check its validity
pub async fn check_token(
    cookie_jar: CookieJar,
) -> Result<(StatusCode, Json<CheckTokenResponse>), (StatusCode, Json<CheckTokenFailedResponse>)> {
    // get token from cookieJar
    let token = cookie_jar
        .get("token")
        .map(|cookie| cookie.value().to_string())
        .ok_or_else(|| {
            (
                StatusCode::BAD_REQUEST,
                Json(CheckTokenFailedResponse {
                    err: "Token is required.".to_string(),
                }),
            )
        })?;

    let user = verify_token(token.as_str()).await;

    match user {
        Ok(_) => Ok((StatusCode::OK, Json(CheckTokenResponse { token }))),
        Err(_) => Err((
            StatusCode::NOT_FOUND,
            Json(CheckTokenFailedResponse {
                err: "user not found".to_string(),
            }),
        )),
    }
}
